07_VirtualFileSystem/02_VirtualFileSystem.md

```diff
@@ -0,0 +1,352 @@
+
+```c
+mountpoints[i].device = device;
+mountpoints[i].tpye = type;
+mountpoints[i].mountpoint = target;
+mountpoints[i[.operations = NULL 
+```
+
+the last line will be populated soon, for now let's leave it to null. 
+
+* The second instruction is for umounting, in this case since we are just unloading the file device from the system, we don't need to know what type is it, so technically we need either the target device or the target folder, the function can actually accept both parameters, but use only one of them, let's call it `vfs_umount`: 
+
+```c
+int vfs_umount(char *device, char *targe);
+```
+
+In this case we need to find the item in the list that contains the required file system, and if found remove it from the list/tree. In our case since we are using an array we need to clear all the fields in the array at the position containing our fs. 
+
+One thing that we should keep in mind is that using an array, once we umount a file system we are just marking that position as available again, so this means that it can be in the middle of the array, and there can be mounted file system after, so we need to keep in mind this while searching for the next free mouuntpoint.
+
+#### A short diversion: the initialization
+
+When the kernel first boot up, of course there is no file system mounted, and all data structures are not allocated, if we decide to use a linked list for example, before the initialization the pointer will point to nothing (or better to garbage) so we will need to allocate the first item of the list to have the first fs accessible. 
+
+In our case since we are using an array we need just to clean all the items in it order to make sure the kernel will not be tricked into thinking that there is a FS loaded, and we need an index pointer to know what is the position of the first available file system.
+
+But where should be the first file system mounted? That again is depending on the project decisions:
+
+* Using a single root approach, the first file system will be mounted on the "/" folder, and this is what we are going to do, this means that all other file systems will be going to stay into subfolders of the root folder. 
+* Using a multi root approach, like windows os, we will have every fs that will have it's own root folder and it will be identified with a letter (A, B, C...)
+* Nothing prevent us to use different approaches, or a mix of them, we can have some file system to share the same root, while some other to have different root, this totally depends on design decision. 
+
+One last thing about the initialization, since our kernel is loaded fully in memory, we don't actually need to have a file system mounted for the kernel to run (even if probably in the future we will need one) so the initialization is just optional to be done during boot time, if the the os already has a kind of shell we can initialize it on the first mount when it will be called. 
```

Comment:
The first part makes sense, but I can't think of a reason to defer initializing the VFS until later, it's super useful to have a filesystem available, even just an initrd. While technically optional, I wouldn't recommend it.

Reply:
mine was a reference to the current naive kernel we are implementing